src/tui/mod.rs

```diff
@@ -15,103 +15,46 @@ pub mod command;
-use state::State;
 use std::sync::atomic::Ordering;
 use std::{io, panic};
 
-/// Representation of a terminal user interface.
+/// Initializes the terminal interface.
 ///
-/// It is responsible for setting up the terminal,
-/// initializing the interface and handling the draw events.
-#[derive(Debug)]
-pub struct Tui<B: Backend> {
-    /// Interface to the Terminal.
-    terminal: Terminal<B>,
-    /// Terminal event handler.
-    pub events: EventHandler,
-    /// Is the interface paused?
-    pub paused: bool,
+/// It enables the raw mode and sets terminal properties.
+pub fn tui_init<B: Backend>(terminal: &mut Terminal<B>) -> Result<()> {
+    terminal::enable_raw_mode()?;
+    ratatui::crossterm::execute!(io::stderr(), EnterAlternateScreen, EnableMouseCapture)?;
+    panic::set_hook(Box::new(move |panic| {
+        better_panic::Settings::auto()
+            .most_recent_first(false)
+            .lineno_suffix(true)
+            .create_panic_handler()(panic);
+        std::process::exit(1);
+    }));
+    terminal.hide_cursor()?;
+    terminal.clear()?;
+    Ok(())
 }
 
-impl<B: Backend> Tui<B> {
-    /// Constructs a new instance of [`Tui`].
-    pub fn new(terminal: Terminal<B>, events: EventHandler) -> Self {
-        Self {
-            terminal,
-            events,
-            paused: false,
-        }
-    }
-
-    /// Initializes the terminal interface.
-    ///
-    /// It enables the raw mode and sets terminal properties.
-    pub fn init(&mut self) -> Result<()> {
-        terminal::enable_raw_mode()?;
-        ratatui::crossterm::execute!(io::stderr(), EnterAlternateScreen, EnableMouseCapture)?;
-        panic::set_hook(Box::new(move |panic| {
-            Self::reset().expect("failed to reset the terminal");
-            better_panic::Settings::auto()
-                .most_recent_first(false)
-                .lineno_suffix(true)
-                .create_panic_handler()(panic);
-            std::process::exit(1);
-        }));
-        self.terminal.hide_cursor()?;
-        self.terminal.clear()?;
-        Ok(())
-    }
-
-    /// [`Draw`] the terminal interface by [`rendering`] the widgets.
-    ///
-    /// [`Draw`]: tui::Terminal::draw
-    /// [`rendering`]: crate::ui:render
-    pub fn draw(&mut self, app: &mut State) -> Result<()> {
-        self.terminal.draw(|frame| ui::render(app, frame))?;
-        Ok(())
-    }
-
-    /// Toggles the [`paused`] state of interface.
-    ///
-    /// It disables the key input and exits the
-    /// terminal interface on pause (and vice-versa).
-    ///
-    /// [`paused`]: Tui::paused
-    pub fn toggle_pause(&mut self) -> Result<()> {
-        self.paused = !self.paused;
-        if self.paused {
-            Self::reset()?;
-        } else {
-            self.init()?;
-        }
-        self.events
-            .key_input_disabled
-            .store(self.paused, Ordering::Relaxed);
-        Ok(())
-    }
+pub fn tui_toggle_pause(paused: &mut bool, events: &EventHandler) {
```

Comment:
Can you expand on this change a bit and explain the rationale?

I honestly don't like the way that this currently works. Instead of taking a mutable reference to a bool, we should handle this in a more stateful way - that's why I had the `Tui` struct before.

Reply:
I honestly don't like the way that this currently works. Instead of taking a mutable reference to a bool, we should handle this in a more stateful way - that's why I had the Tui struct before.

Looks like we can remove `paused` variable and simply lock user input at `Event::Trace` event, and unlock at `Event::TraceResult`.